### PR TITLE
Trying to make peace with adamw over locate_asset

### PR DESCRIPTION
**Pull request by coolo**
Instead of returning self constructed URLs from helper functions, return
a relative path passed to URL helpers in the controller.

This way we don't hard code the /fixed/ part anywhere but the utility
function and still use URL helpers

### CODE REVIEW

File: lib/OpenQA/Schema/Result/Assets.pm [hunk 1]
```diff
@@ -15,18 +15,17 @@
 # 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
 package OpenQA::Schema::Result::Assets;
-use base qw/DBIx::Class::Core/;
+use base qw(DBIx::Class::Core);
```
Comment (coolo) [inline lib/OpenQA/Schema/Result/Assets.pm]: I figured that. but it's a function not an operator

Comment (aaannz) [inline lib/OpenQA/Schema/Result/Assets.pm]: But .. but even perldoc for `qw` uses slashes!

Comment (coolo) [inline lib/OpenQA/Schema/Result/Assets.pm]: in some places:

coolo@cleopatra#os-autoinst>man perlfunc | grep 'qw(' | wc -l
19
coolo@cleopatra#os-autoinst>man perlfunc | grep 'qw/' | wc -l
2

File: lib/OpenQA/Utils.pm [hunk 2]
```diff
@@ -332,18 +332,24 @@ sub parse_assets_from_settings {
...
 }
 
+sub _relative_or_absolute {
+    my ($path, $relative) = @_;
+
+    return $path if $relative;
+    return catfile($assetdir, $path);
+}
+
 # find the actual disk location of a given asset.
```
Comment (AdamWill) [inline lib/OpenQA/Utils.pm]: this needs updating now.

File: t/15-assets.t [hunk 3]
```diff
@@ -182,16 +185,19 @@ ok($fixed->is_fixed(), 'fixed should be considered a fixed asset');
...
-$expected = catfile($OpenQA::Utils::assetdir, 'repo', 'testrepo');
-is(locate_asset('repo', 'testrepo', 1), $expected, 'locate_asset should find testrepo in expected location');
+$expected = catfile($OpenQA::Utils::assetdir, 'repo', 'tmprepo');
+is(locate_asset('repo', 'tmprepo', mustexist => 1), $expected, 'locate_asset should find tmprepo in expected location');
 
 # non-existent ISO asset
 $expected = catfile($OpenQA::Utils::assetdir, 'iso', 'nex.iso');
-is(locate_asset('iso', 'nex.iso', 0), $expected, 'locate_asset 0 should give location for non-existent asset');
-ok(!locate_asset('iso', 'nex.iso', 1), 'locate_asset 1 should not give location for non-existent asset');
+is(locate_asset('iso', 'nex.iso'), $expected, 'locate_asset 0 should give location for non-existent asset');
```
Comment (AdamWill) [inline t/15-assets.t]: well, this was intended to be a test of what happens when you explicitly set `0`, not a test of 'whatever the default currently is'. Of course to be truly comprehensive we would test both, I guess.

File: t/api/02-assets.t [hunk 4]
```diff
@@ -84,14 +84,14 @@ for my $i ($iso1, $iso2) {
 
 my $listing = [
     {
-        id       => 6,
+        id       => 7,
```
Comment (AdamWill) [inline t/api/02-assets.t]: why?

File: t/data/openqa/factory/repo/.gitignore [hunk 5]
```diff
@@ -1,4 +0,0 @@
-# Ignore everything in this directory
```
Comment (AdamWill) [inline t/data/openqa/factory/repo/.gitignore]: why delete this? so people will notice if the test process doesn't clean up after itself?

File: t/fixtures/01-jobs.pl [hunk 6]
```diff
@@ -234,7 +239,7 @@
         VERSION   => '13.1',
         MACHINE   => '32bit',
         backend   => 'qemu',
-        jobs_assets => [{asset_id => 1}, {asset_id => 5}],
+        jobs_assets => [{asset_id => 1}, {asset_id => 5},],
```
Comment (AdamWill) [inline t/fixtures/01-jobs.pl]: why?

File: t/fixtures/01-jobs.pl [hunk 7]
```diff
@@ -378,16 +383,18 @@
         state      => "running",
         t_finished => undef,
         backend    => 'qemu',
-        t_started  => time2str('%Y-%m-%d %H:%M:%S', time - 600, 'UTC'),     # 10 minutes ago
-        t_created  => time2str('%Y-%m-%d %H:%M:%S', time - 7200, 'UTC'),    # Two hours ago
```
Comment (AdamWill) [inline t/fixtures/01-jobs.pl]: why?

**Comment by AdamWill** [comment]
So I added several notes, but I actually really like this, thanks a lot - better than my ideas. However, can you please at least separate all the 'cleanups' from the actual change? And can you reconsider the cleanups which aren't really obviously better but seem more like personal tics? I have a specific reason why: I use `git blame` a lot, and it's really annoying when you're trying to find a change with it, but you keep hitting little fiddly 'cleanups' instead.

At least if you're going to change the uses of `qw`, change *all uses of it in openQA in a single commit*, and make it a separate PR or something, because it really has nothing to do with this fix.

    [adamw@adam openQA (fixed-asset-download %)]$ grep -R "qw */" | wc -l
    309

But this is great! Don't mean to sound ungrateful. :)

File: t/api/02-assets.t (hunk 4 above)
Comment (coolo) [inline t/api/02-assets.t]: Because I added one to the fixtures, the next id has to be higher. Fixtures are crap :(

File: t/data/openqa/factory/repo/.gitignore (hunk 5 above)
Comment (coolo) [inline t/data/openqa/factory/repo/.gitignore]: Because I added a README to testrepo

File: t/fixtures/01-jobs.pl (hunk 6 above)
Comment (coolo) [inline t/fixtures/01-jobs.pl]: aehm, this is a leftover when I tried to add asset 6 here - but that caused too many unrelated tests to fail, so I removed it again. It won't hurt though

File: t/fixtures/01-jobs.pl (hunk 7 above)
Comment (coolo) [inline t/fixtures/01-jobs.pl]: why I moved comments? to fight tidy

**Comment by coolo** [comment]
So guys, I did as you wanted. I split the commits and left my fight for qw function for later.

File: lib/OpenQA/WebAPI/Controller/File.pm [hunk 8]
```diff
@@ -108,17 +109,12 @@ sub test_asset {
...
-    $path = $self->url_for($urlname, assettype => $asset{type}, assetname => $asset{name});
-    if ($self->param('subpath')) {
-        $path .= '/' . $self->param('subpath');
-        # better safe than sorry. Mojo seems to canonicalize the
-        # urls for us already so this is actually not needed
-        return $self->render_exception('invalid character in path') if ($path =~ /\/\.\./ || $path =~ /\.\.\//);
-    }
+    my $path = locate_asset($asset{type}, $asset{name}, relative => 1);
+    $path = catfile($path, $self->param('subpath')) if $self->param('subpath');
+    return $self->render(text => 'invalid character in path', status => 400) if ($path =~ /\/\.\./ || $path =~ /\.\.\//);
```
Comment (AdamWill) [inline lib/OpenQA/WebAPI/Controller/File.pm]: It's a nitpick, and to follow my own rules should maybe be a separate commit, but this message is dumb. It's not about an 'invalid character' at all. The error message should reflect what it's *actually* disallowing.

**Comment by AdamWill** [comment]
Thanks a lot! Haven't run it yet, but definitely I'm an eyeball +1. I wouldn't object if you wanna merge it, since the tests pass.

File: lib/OpenQA/WebAPI/Controller/File.pm (hunk 8 above)
Comment (coolo) [inline lib/OpenQA/WebAPI/Controller/File.pm]: I only changed the if not the message.

Comment (AdamWill) [inline lib/OpenQA/WebAPI/Controller/File.pm]: yeah, I know - it's just something that's been bugging me all along, so I figured I'd mention it :)
